From a5744177a4c381a105d11e0563d850467f7e2390 Mon Sep 17 00:00:00 2001
From: Yang Shi <yang.shi@windriver.com>
Date: Fri, 11 Nov 2011 14:38:08 -0800
Subject: [PATCH 148/238] MIPS: No branches in delay slots for huge pages in handle_tlbl

Source: Cavium SDK 2.2-414

For the case PM_DEFAULT_MASK == 0, we were placing a branch in the
delay slot of another branch.  This leads to undefined behavior.

Signed-off-by: David Daney <david.daney@cavium.com>
Integrated-by: Yang Shi <yang.shi@windriver.com>
---
 arch/mips/mm/tlbex.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/arch/mips/mm/tlbex.c b/arch/mips/mm/tlbex.c
index a298e10..fa9026a 100644
--- a/arch/mips/mm/tlbex.c
+++ b/arch/mips/mm/tlbex.c
@@ -1974,6 +1974,8 @@ static void __cpuinit build_r4000_tlb_load_handler(void)
 		/* Reload the PTE value */
 		iPTE_LW(&p, wr.r1, wr.r2); 
 
+		if (PM_DEFAULT_MASK == 0)
+			uasm_i_nop(&p);
 		/*
 		 * We clobbered C0_PAGEMASK, restore it.  On the other branch
 		 * it is restored in build_huge_tlb_write_entry.
-- 
1.7.0

